nexus/src/app/sagas/instance_start.rs

```diff
@@ -129,9 +133,13 @@ impl NexusSaga for SagaInstanceStart {
     }
 
     fn make_saga_dag(
-        _params: &Self::Params,
+        params: &Self::Params,
         mut builder: steno::DagBuilder,
     ) -> Result<steno::Dag, SagaInitError> {
+        if params.reason == Reason::AutoRestart {
+            builder.append(set_auto_restart_timestamp_action());
+        }
```

Comment:
Related to the comment down on 196-ish: I wonder if it would make more conceptual sense to have this action happen in the reincarnation task and not in the start saga proper. The thing we're doing--saying "yeah I tried to automatically restart this instance"--seems proper to that task; this step doesn't have an undo action; and its position in the DAG means we're not relying on the synchronization properties that come with doing it later in the saga (i.e. having locked out other attempts to start the instance).

I don't think this is a functional problem, to be clear; just a question of what makes the reincarnation task and the start saga more readable. WDYT?

Reply:
Hmm, so, my thought process was that it was important to not set the timestamp until we were *sure* that the saga was actually started and existed in the database. I didn't want to have a situation where the background task sets the timestamp and then the Nexus process crashes before the saga is actually started, making other Nexii  no longer see that instance as needing a start saga when it actually still needs one. 

But, it occurs to me that the `StartSaga` trait now allows us to await the saga having started separately from awaiting its completion, so we could just wait until the saga is started in the background task, update the timestamp, and then wait for it to complete. That would let us get rid of the extra complexity in the start  saga. But, in that case, it might behoove us to _not_ have the auto-restart timestamp be guarded  by the instance's state  generation, and instead have a special  query to just  set it conditional on the new timestamp  being  >= the previous one. That  way, we can't get into a situation where we start a saga, go to set the timestamp, and can't because the generation  number has changed.

Does that seem reasonable to you?

Reply:
We discussed this for a while offline. To summarize the points we covered:

1. Trying to set the last-restarted time in the restarter task (as opposed to the start saga) runs the risk of a race in which a long-delayed restarter task observes that an instance is Failed when selecting instances to try to restart, then gets delayed for a long time before actually trying to restart it, during which time the instance is restarted by some other means. This pushes out the instance's cooldown if it fails again.
2. Writing the query as "only set the last-restart time if the instance is still Failed" helps, but only if the restarter task tries to set the last-restart time before actually kicking off a start saga. (If the start saga is running, it may move the instance out of Failed before the restarter task gets to set the time.) And waiting until the saga is actually started runs into the problem @hawkw identified above, where the restart time gets written and then the task crashes, such that no restart was actually attempted.
3. As a general matter, trying to set the last-restart timestamp before getting to `sis_move_to_starting` is perilous, because multiple restarter tasks/start sagas are allowed to run simultaneously before that point, and making sure they don't clobber each other's updates is hard. The flipside of waiting for `move_to_starting` to set the timestamp is that there are some possibly non-transient reasons a start saga could fail (e.g. resource exhaustion), and if the saga hits any of these, the instance will return to the Failed state and not be removed from the set of instances requiring a restart. (This makes it hard to use the `ORDER BY random() LIMIT n` approach to picking instances to restart: that method generally assumes that once an instance is picked, the picker can prevent it from being picked again.)

One way to deal with all of these problems is to add another Instance column (sigh) that stores the ID of the last task to try to auto-restart an instance. This allows an individual restarter task to mark instances that it has seen before without worrying about whether the start saga got far enough to put the instance back on cooldown. (It also does this without having to add a possibly-ever-expanding `NOT IN` column to the "find target instances" query.)

Another alternative is abandon the "pick N random instances" approach and go back to paginating, but that has the problems described in #6565 (restarter tasks running in phase with each other will run useless duplicative start sagas).

Giving the reincarnation task a way to use the DB to mark previously-seen instances seems better to me than just giving up and paginating once more, but I have some FUD about adding another Instance column (and presumably an index thereon) just for this purpose; most of the time, no restarter is running, and instances are hopefully not going to go to Failed very often, so the column isn't going to be useful most of the time. I think that, if it were totally up to me, I would rank the options here as

1. find some (inexpensive) way to deal with #6565 that avoids the same-instance-selected-twice problem by other means
2. go back to paginating and figure out how to deal with #6565 only when we've actually seen it become a problem
3. add database state to make the `ORDER BY`/`LIMIT` trick work well for this case

Reply:
I think for now, I'm probably going to just change this back to paginating the query, since I can't really come up with any obvious way to make the same-instance-selected-twice problem go away here. We can just deal with #6565 later (e.g. by adding an additional UUID column) if it becomes a problem in real life. AFAICT, it shouldn't be a correctness issue, just efficiency --- and it means that there will be a lot of "failed" sagas that might raise flags when debugging but will turn out to be totally innocuous.